### ISSUE TITLE
FR: `--exit-node-allow-lan-access=192.168.0.0/16` to allow routed local subnets

### ISSUE THREAD

**Issue by 0x5c** [tailscale/tailscale#3646]
### What is the issue?

When an exit node is being used, the computer fails to access anything on other local subnets.

For exemple, if connected to the subnet 192.168.0.1/24, reaching the subnet 192.168.1.1/24 is impossible once an exit node is being used.

It seems like tailscale has no way to detect this and simply declares anything outside the subnet as being the internet.

### Steps to reproduce

1. Have a router setup with multiple subnets, nat, and internet access.
2. Connect your computer to one subnet
3. Enable usage of an exit node (with the flag to allow local traffic)
4. All traffic to the other subnets is now blackholed to oblivion, no replies are ever received.

### Are there any recent changes that introduced the issue?

Not that I'm aware, the problem is new to me, as is the network setup that has the subnets.

### OS

Linux

### OS version

Voidlinux (rolling distro)

### Tailscale version

1.18.2

### Bug report

BUG-ec3d643e55e290b2e586427f959dad1768fcb5bfa08310b43140bd3b35f6b5cc-20220102022640Z-23bff9456e38cb53

**Comment by DentonGentry** [comment]
This is correct as described: the Allow LAN Access feature allows the specific subnet the device is on, not any further subnets which can be reached via local routing. I understand the intent of the feature request: everything in my corporate net / homelab / etc is safe and should be allowed while using an exit node, but anything on the public Internet is not safe and should use the exit node.

Unfortunately I don't think we could apply that as a policy in all cases, to allow 192.168.0.0/16 + 10.0.0.0/8 + 172.16. 0.0/12 when Allow LAN Access is set. I'm not sure we can reasonably handle complex local network topologies when using the exit node feature.

**Comment by bradfitz** [comment]
Maybe in the future we could have some mode where, say, 100.100.100.100 can be a next hop for the exit node you've chosen to use but then you can go and configure your routes manually.

Or maybe we could just extend `--exit-node-allow-lan-access` or something like it to take a list of route and next hops to carve out, but that's a little niche.

**Comment by 0x5c** [comment]
Yeah, now that I've thought about it more it doesn't seem to me like it could be automated in any way.

But for my use case a manual setting would be perfectly fine, so I'd like to turn this into a feature request.

I can see maybe something like `--exit-node-allow-lan-access='192.168.0.0/16'`

**Comment by Omrigan** [comment]
As suggested in #1527 for the time being we can use one of two options

----
For now I succesfully do it manually like this `sudo ip route add throw 192.168.0.0/16 table 52`.

_Originally posted by @boothb in https://github.com/tailscale/tailscale/issues/1527#issuecomment-1003060688_

----

As an aside for anyone reading this, I did discover an alternative that gave me the same thing I wanted.

`tailscale up --advertise-exit-node --advertise-routes=192.168.0.0/16`

Now I have that, I'm not even sure how a client-side setting would interact with this option.

_Originally posted by @marcosscriven in https://github.com/tailscale/tailscale/issues/1527#issuecomment-1121923535_

**Comment by Omrigan** [comment]
By the way, why don't we just ignore all three private network ranges by default for something like this?

```
$ ip route list table 52
default dev tailscale0
100.73.246.34 dev tailscale0
100.83.242.16 dev tailscale0
100.89.23.12 dev tailscale0
100.100.100.100 dev tailscale0
100.127.159.75 dev tailscale0
throw 127.0.0.0/8
throw 172.16.0.0/12
throw 192.168.0.0/16
throw 10.0.0.0/8
```

**Comment by DentonGentry** [comment]
> By the way, why don't we just ignore all three private network ranges by default for something like this?

I believe that would break interaction with those subnets being advertised as routes by other nodes on the tailnet.

----

I think the original request is best met with: `tailscale up --advertise-exit-node --advertise-routes=192.168.0.0/16`

**Comment by 0x5c** [comment]
> I think the original request is best met with: `tailscale up --advertise-exit-node --advertise-routes=192.168.0.0/16`

@DentonGentry I do not think this does what the request is about.

This request is about accessing other local subnets *on the client*, not on the exit node.

The use-case is a machine (local ip `192.168.0.100/24`) that uses an exit node, but needs access to the whole `/16` local network. Access to the local network of the client must not depend on what exit node is being used.

Here it seems your solution exposes the exit node's local network, which would even break the basic `--exit-node-allow-lan-access` usage

**Comment by bjammin** [comment]
I would love to see this feature properly supported on all clients.
Exit nodes are basically useless for me as I can't  access other local subnets while connected.

**Comment by amosbird** [comment]
> I believe that would break interaction with those subnets being advertised as routes by other nodes on the tailnet.

If `exit-node-allow-lan-access='192.168.0.0/16'` is available, clients have full control over which subnets they choose to use. We can still keep the original behavior unchanged if no value is given.

**Comment by swannysec** [comment]
Just chiming in to say this issue is a significant blocker for one of my use-cases. The solution noted above seems elegant and of low overall risk to me.  Is there any additional data we can provide to help prioritize this issue?

**Comment by sandydoo** [comment]
I run OrbStack and a bunch of Linux VMs on my macOS machine. Each runs on its own local subnet with a bridge and vmnet.

When I connect to an exit node (with `--exit-node-allow-lan-access` enabled), these local IPs are routed over the default interface and out into the internet.

Here, the VM bridge 192.168.64.0/24 is being routed to the router after enabling the exit node.
```
Internet:
Destination        Gateway            Flags               Netif Expire
192.168.64         192.168.100.1      UGSc                  en0
```

After disabling the exit node, Orbstack somehow recovers (reapplies the routes?). The rest require manual intervention:
```
sudo route change -net 192.168.64/24 -interface bridge100
```

**Comment by Ramblurr** [comment]
I just paid for a month of tailscale's mullvad addon, and planned to switch to it from normal mullvad. But then I ran into this issue, which totally breaks the use case of:

"`Use tailscale exit nodes (maybe a mullvad node) for everything except my entire office/home which is more than one subnet`"

Prior art in solving this issue:

* Vanilla mullvad offers split tunneling via linux kernel connection tracking and meta marks [source](https://mullvad.net/en/help/split-tunneling-with-linux-advanced#allow-ip)

    This would be an elegant solution because it does not expose complicated feature flags to the tailscale UX that also may not work on a platform. It just requires internal changes to tailscale's nftables config and then the user has to do the work of excluding things themselves.

* Tailscale's Android app itself supports split tunneling, which sets a precedent for platform-specific tailscale features

**Comment by 0x5c** [comment]
@Ramblurr this is already a feature request for the ability to specify the size of a local network, something tailscale cannot find by itself for multi-subnet local networks.
The flag already exists for enabling access to the local network, is the only way to do so, and would be trivial to extend such that the os' routing tables also include direct access to the other local subnets.

Metamarks handling seems more complex to implement both for the program and the user, but also seems like something that should be in a separate feature request.

**Comment by Ramblurr** [comment]
thanks for the reply @0x5c

> this is already a feature request for the ability to specify the size of a local network, something tailscale cannot find by itself for multi-subnet local networks.

Do by size you mean "all the valid subnets that should be considered local"?  For example, my office has multiple vlans with various subnets that are *not* contiguous.

I would like to be able to pass `--exit-node-allow-lan-access=192.168.0.0/24,10.9.8.0/22,10.9.4.0/22` ..etc

That seems on-top for this issue?

>  The flag already exists for enabling access to the local network, is the only way to do so, and would be trivial to extend such that the os' routing tables also include direct access to the other local subnets.

Totally agreed!

> Metamarks handling seems more complex to implement both for the program and the user, but also seems like something that should be in a separate feature request.

Oh I agree that it is more complex. I only suggested it as an alternative to the proposed `--exit-node-allow-lan-access=...` because there seems to be some concern from tailscale developers about cluttering the UX of the tailscale cli and their desire to not handle "complex local network topologies".

Adding connection tracking handling  ([meta mark is already there!](https://github.com/tailscale/tailscale/blob/main/util/linuxfw/linuxfw.go#L87)) would solve this issue at a layer that does not complicate UX, hence my suggestion. (Incidentally it would also allow `mullvad-exclude other-program` type of split-tunneling, but that's definitely outside the scope of this issue.)

I would be satisfied with `--exit-node-allow-lan-access=192.168.0.0/24,10.9.8.0/22,10.9.4.0/22`

**Comment by 0x5c** [comment]
Ah I see, yeah the notation in the title only allows for contiguous subnets. The notation you propose seems like a logical extension of the idea, and also fairly simple to handle

**Comment by puigru** [comment]
This is a blocker for me. I use RDP to connect to another machine running tailscale, they're both in different networks. My host is in 192.168.1.0/24 and the one I'm remoting into that has tailscale is in 192.168.3.0/24. As soon as I set an exit node, my RDP session dies.
This was quite a head scratcher for me at first, I thought the firewall rules in my router were misconfigured but it was this issue all along. I suppose I expected that "Allow LAN access" meant the whole 192.168.0.0/16 network was accessible on the source network.

**Comment by bjammin** [comment]
This is still a problem for me too.

Was really easy when we were using OpenVPN and I'm surprised it's not been added as an option to Tailscale yet.

**Comment by aradng** [comment]
> This is a blocker for me. I use RDP to connect to another machine running tailscale, they're both in different networks. My host is in 192.168.1.0/24 and the one I'm remoting into that has tailscale is in 192.168.3.0/24. As soon as I set an exit node, my RDP session dies. This was quite a head scratcher for me at first, I thought the firewall rules in my router were misconfigured as I was trying to isolate this machine, but it was this issue all along. I suppose I expected that "Allow LAN access" meant the whole 192.168.0.0/16 network was accessible on the source network.

as a tmp fix you can add static routes to your main routing table (either using ip or netplan)
```bash
ip route add 192.168.0.0/16 dev eth0
```
you can add arbitrary routing in main table for your local networks and tailscalle will respect it as long as it's not using the default route implicitly. e.g. `ip route add 172.16.0.0/16 via 10.0.0.1 dev ens18`

**Comment by valarauca** [comment]
👋 just chiming in to say this featured is **required**. 

My use case is (I'll grant you a little weird). 1 home fan LAN has JUMBO frames (and 50Gbe) the the other (slow LAN) connects to the internet and has wifi. I setup an exit node and the whole fast LAN goes dark. Now, I fully understand using the APIPA CIDR is non-standard (169.254.0.0/16), but for a LAN that doesn't need internet access it works A-OK. 

---

Thanks for everyone who mentioned the `ip route add` fix. I spent hours digging through logs before finding this issue.

**Comment by raesene** [comment]
Just to say I'm another impacted user. My use case is using Tailscale + Mullvad and I wanted to create ephemeral browser containers using webtop, buut I can't access the webtop container when I have an exit node configured. I can see some workarounds in this thread, which is great but it'd be really nice to have an option to specify which networks are considered local.

**Comment by coxde** [comment]
Hi I also want to point out this would be a MUST feature. I just spent hours to troubleshooting my network, turns out it's because Tailscale doesn't recognize other VLANs as the LAN. It'd be really nice to have such options instead of manually doing ip route.

**Comment by segment-tree** [comment]
Any progress about that?

**Comment by MacsInSpace** [comment]
Adding my +1 here. Thanks for the temp fix. 
Currently doing:
`sudo route add -net 10.0.0.0/8 $GW`

This would be a great idea (mentioned earlier)
`-exit-node-allow-lan-access='10.0.0.0/8'`
Thanks for the great product!
